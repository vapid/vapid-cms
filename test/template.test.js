const { readFileSync } = require('fs');
const { resolve } = require('path');
const TemplateCompiler = require('../lib/TemplateCompiler');

/**
 * Constructor
 */
describe('.constructor', () => {
  test('replaces partial with file contents', () => {
    const partials = [resolve(__dirname, 'fixtures', '_partial.html')];
    const partialContent = readFileSync(partials[0], 'utf-8');
    const template = new TemplateCompiler('{{> partial}}', { partial: partialContent });

    expect(template.html).toEqual(partialContent);
  });

  test('allows partials within partials', () => {
    const partials = {
      outer: '{{> inner}}',
      inner: 'inner content',
    };
    const template = new TemplateCompiler('{{> outer}}', partials);

    expect(template.html).toEqual(partials.inner);
  });
});

/**
 * TemplateCompiler.fromFile
 */
describe('.fromFile', () => {
  test('creates new instance from a file path', () => {
    const filePath = resolve(__dirname, 'fixtures/basic.html');
    const template = TemplateCompiler.fromFile(filePath);
    const html = readFileSync(filePath, 'utf-8');

    expect(html).toEqual(template.html);
  });

  test('reads partials from disk', () => {
    const basePath = resolve(__dirname, 'fixtures');
    const filePath = resolve(basePath, 'with_partials.html');
    const partials = [
      resolve(basePath, '_partial.html'),
      resolve(basePath, 'subdirectory/_partial.html'),
    ];
    const template = TemplateCompiler.fromFile(filePath, basePath, partials);

    expect(template.render()).toMatchSnapshot();
  });
});

/**
 * template.parse
 */
describe('#parse', () => {
  test('puts fields into separate branches', () => {
    const html = `
      {{name}}
      {{#section about}}
        {{name}}
        {{bio}}
      {{/section}}`;
    const tree = new TemplateCompiler(html).parse();

    expect(tree).toMatchSnapshot();
  });

  test('allow sections without keyword', () => {
    const withKeyword = new TemplateCompiler('{{#section about}}{{/section}}').parse();
    const withoutKeyword = new TemplateCompiler('{{#about}}{{/about}}').parse();

    expect(withKeyword['section about'].name).toEqual(withoutKeyword.about.name);
  });

  /* eslint-disable quotes */
  test('parses parameters with optional quotes', () => {
    const plain = new TemplateCompiler('{{test required=false}}').parse();
    const single = new TemplateCompiler(`{{test required='false'}}`).parse();
    const double = new TemplateCompiler('{{test required="false"}}').parse();

    expect(plain.general.fields['test required=false'].params).toEqual(single.general.fields[`test required='false'`].params);
    expect(plain.general.fields['test required=false'].params).toEqual(double.general.fields['test required="false"'].params);
  });

  test('parses parameters with escaped quotes', () => {
    const inTheMiddle = new TemplateCompiler(`{{test placeholder="Testing \\"quotes\\" in the middle."}}`).parse();
    const mixed = new TemplateCompiler(`{{test placeholder="If this'll work"}}`).parse();

    expect(inTheMiddle).toMatchSnapshot();
    expect(mixed).toMatchSnapshot();
  });
  /* eslint-enable quotes */

  test('parses order by clause', () => {
    const orderBy = new TemplateCompiler('{{#section offices order=city,-name}}{{/section}}').parse();
    expect(orderBy).toMatchSnapshot();
  });

  test('parses conditionals', () => {
    const conditional = '{{#if foo}}Foo{{/if}}';

    expect(new TemplateCompiler(conditional).parse()).toMatchSnapshot();
    expect(new TemplateCompiler(
      conditional,
      {},
    ).parse()).toMatchSnapshot();
  });

  test('parses general references in sections correctly', () => {
    const template = new TemplateCompiler(`
      {{test}}
      {{#section child}}{{general.test}}{{/section}}
    `).parse();

    expect(Object.keys(template.general.fields)).toEqual(['test']);
    expect(Object.keys(template['section child'].fields)).toEqual(['general.test']);
  });
});

/**
 * template.render
 */
describe('#render', () => {
  test('replaces tags with content', () => {
    const template = new TemplateCompiler('Hello, {{name}}.');
    const rendered = template.render({ general: { name: 'World' } });

    expect(rendered).toMatchSnapshot();
  });

  test('replaces general references in sections with content', () => {
    const template = new TemplateCompiler(`
      {{greeting}} General
      {{#section child}}{{general.greeting}} Section{{/section}}
    `);
    const rendered = template.render({
      general: {
        greeting: 'Hello',
      },
      'section child': {
        'general.greeting': 'Hello',
      },
    });

    expect(rendered).toMatchSnapshot();
  });
});
