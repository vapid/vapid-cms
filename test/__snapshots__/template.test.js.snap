// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#parse parses conditionals 1`] = `
Object {
  "general": Object {
    "fields": Object {
      "foo": Object {
        "context": null,
        "key": "foo",
        "params": Object {},
        "path": "foo",
        "remainder": "",
      },
    },
    "keyword": undefined,
    "name": "general",
    "params": Object {},
  },
}
`;

exports[`#parse parses conditionals 2`] = `
Object {
  "general": Object {
    "fields": Object {
      "foo": Object {
        "context": null,
        "key": "foo",
        "params": Object {},
        "path": "foo",
        "remainder": "",
      },
    },
    "keyword": undefined,
    "name": "general",
    "params": Object {},
  },
}
`;

exports[`#parse parses order by clause 1`] = `
Object {
  "general": Object {
    "fields": Object {},
    "keyword": undefined,
    "name": "general",
    "params": Object {},
  },
  "section offices order=city,-name": Object {
    "fields": Object {},
    "keyword": "section",
    "name": "offices",
    "params": Object {
      "order": "city,-name",
    },
  },
}
`;

exports[`#parse parses parameters with escaped quotes 1`] = `
Object {
  "general": Object {
    "fields": Object {
      "test placeholder=\\"Testing \\\\\\"quotes\\\\\\" in the middle.\\"": Object {
        "context": null,
        "key": "test",
        "params": Object {
          "placeholder": "Testing \\"quotes\\" in the middle.",
        },
        "path": "test",
        "remainder": "placeholder=\\"Testing \\\\\\"quotes\\\\\\" in the middle.\\"",
      },
    },
    "keyword": undefined,
    "name": "general",
    "params": Object {},
  },
}
`;

exports[`#parse parses parameters with escaped quotes 2`] = `
Object {
  "general": Object {
    "fields": Object {
      "test placeholder=\\"If this'll work\\"": Object {
        "context": null,
        "key": "test",
        "params": Object {
          "placeholder": "If this'll work",
        },
        "path": "test",
        "remainder": "placeholder=\\"If this'll work\\"",
      },
    },
    "keyword": undefined,
    "name": "general",
    "params": Object {},
  },
}
`;

exports[`#parse puts fields into separate branches 1`] = `
Object {
  "general": Object {
    "fields": Object {
      "name": Object {
        "context": null,
        "key": "name",
        "params": Object {},
        "path": "name",
        "remainder": "",
      },
    },
    "keyword": undefined,
    "name": "general",
    "params": Object {},
  },
  "section about": Object {
    "fields": Object {
      "bio": Object {
        "context": null,
        "key": "bio",
        "params": Object {},
        "path": "bio",
        "remainder": "",
      },
      "name": Object {
        "context": null,
        "key": "name",
        "params": Object {},
        "path": "name",
        "remainder": "",
      },
    },
    "keyword": "section",
    "name": "about",
    "params": Object {},
  },
}
`;

exports[`#render replaces general references in sections with content 1`] = `
"      Hello General
      Hello Section
    "
`;

exports[`#render replaces tags with content 1`] = `"Hello, World."`;

exports[`.fromFile reads partials from disk 1`] = `
"<!DOCTYPE html>
<html>
<head>
  <title></title>
</head>
<body>
  Hello, .

  I'm a partial. 
  I'm a subdirectory partial. 
</body>
</html>"
`;
